extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayBuildStep.java

```diff
@@ -0,0 +1,110 @@
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.ExecutionTime;
+import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.substrate.SubstrateResourceBuildItem;
+
+class FlywayBuildStep {
+
+    private static final String FLYWAY_DATABASES_PATH_ROOT = "org/flywaydb/core/internal/database";
+    private static final String FLYWAY_METADATA_TABLE_FILENAME = "createMetaDataTable.sql";
+    private static final String[] FLYWAY_DATABASES_WITH_SQL_FILE = {
+            "cockroachdb",
+            "derby",
+            "h2",
+            "hsqldb",
+            "mysql",
+            "oracle",
+            "postgresql",
+            "redshift",
+            "saphana",
+            "sqlite",
+            "sybasease"
+    };
+
+    private static final Logger LOGGER = Logger.getLogger(FlywayBuildStep.class);
+
+    @BuildStep
+    FeatureBuildItem feature() {
+        return new FeatureBuildItem(FeatureBuildItem.FLYWAY);
+    }
+
+    @Record(ExecutionTime.STATIC_INIT)
+    @BuildStep(providesCapabilities = "io.quarkus.flyway")
+    void registerSubstrateResources(
+            BuildProducer<SubstrateResourceBuildItem> resource,
+            ApplicationArchivesBuildItem appArchives) throws IOException, URISyntaxException {
+        Path root = appArchives.getRootArchive().getArchiveRoot();
+        List<String> resources = generateDatabasesSQLFiles();
+        resources.addAll(discoverApplicationMigrations(root));
+        resource.produce(new SubstrateResourceBuildItem(resources.toArray(new String[0])));
+    }
+
+    private List<String> discoverApplicationMigrations(Path root) throws IOException, URISyntaxException {
+        List<String> resources = new ArrayList<>();
+        try {
+            // TODO: this should be configurable, using flyway default
+            String location = "db/migration";
+            Enumeration<URL> migrations = Thread.currentThread().getContextClassLoader().getResources(location);
+            while (migrations.hasMoreElements()) {
+                URL path = migrations.nextElement();
+                LOGGER.info("Adding application migrations in path: " + path);
+                List<String> applicationMigrations = walk(Paths.get(path.toURI()))
+                        .filter(Files::isRegularFile)
+                        .map(it -> Paths.get(location, it.getFileName().toString()).toString())
+                        .peek(it -> LOGGER.info("Discovered: " + it))
+                        .collect(Collectors.toList());
+                resources.addAll(applicationMigrations);
+            }
+            return resources;
+        } catch (IOException | URISyntaxException e) {
+            LOGGER.fatal("Error discovering application migrations: " + e.getMessage(), e);
```

Comment:
In general code should either log or re-throw, but not both. This should be just throwing a new exception this the message, and it will get logged by the normal error handing code.